### ISSUE TITLE
How to get posterior error, spectrum q, peptide q, protein q?

### ISSUE THREAD

**Issue by daichengxin** [theGreatHerrLebert/sagepy#13]
Hi @theGreatHerrLebert , Thansks for your great work! I followed the README documentation but got PSM all scored as 1. So, how do I calculate and get the posterior error, spectrum q, peptide q, protein q values?

**Comment by theGreatHerrLebert** [comment]
Hey there,

Thanks a lot! 😊 I admit that the documentation is still lagging behind, partly because there are ongoing changes that are affecting the current API for SAGE.

1. For **spectrum-q** and **peptide-q**, there is the [target_decoy_competition](https://github.com/theGreatHerrLebert/sagepy/blob/main/sagepy/sagepy/qfdr/tdc.py#L28), where you can choose between different versions of target-decoy competition to calculate q-values from raw scores. I’ve also added a [notebook example](https://github.com/theGreatHerrLebert/sagepy/blob/main/sagepy/examples/scoring/scoring.ipynb) that demonstrates how to perform q-value calculation for a collection of identified spectra.

2. I’ve added **posterior error probability (PEP)** as a standalone function in a current [feature branch](https://github.com/theGreatHerrLebert/sagepy/blob/feature/PEP/sagepy/sagepy/core/ml/pep.py#L125). The original SAGE implementation was straightforward enough to re-implement in pure Python rather than wrapping the original code. However, this branch contains some breaking changes, as I’m also working on an easier way to pass PTMs—one of my long-standing goals to simplify. I plan to merge all currently developed features at the end of the week , #14, and create a new PyPI release.

3. Protein-q is not implemented at this point, as it requires additional consideration on how protein groups should be inferred from observed peptides. I need to review how SAGE handles this and potentially add some method options, given the extensive literature and ongoing debate on the best approach.

I hope this helps!

Best regards,
David

**Comment by daichengxin** [comment]
Thanks for your rapid response. Looking forward to the new PyPI release😀

**Comment by theGreatHerrLebert** [comment]
Hey @daichengxin,

I just updated the PyPI package to contain the latest changes, which also includes a  function to calculate the PEP from a given set of scores and labels. I will keep the issue open until the protein-q is added. Usage example can be found [here](https://github.com/theGreatHerrLebert/sagepy/blob/main/sagepy/examples/scoring/scoring.ipynb). Let me know if something does not work as expected :) 

Best,

David

**Comment by daichengxin** [comment]
Hi @theGreatHerrLebert, the latest version works! Thanks a lot. I also encountered another problem. It reported an error as follows when I use rescore_lda function.  Then I printed a PeptideSpectrumMatch, the fragments_observed values are indeed empty.
![image](https://github.com/user-attachments/assets/81f6436e-8a66-4c2c-9622-91640f2df106)
![image](https://github.com/user-attachments/assets/38a473f7-72db-47a5-b0b3-53ad57677f0b)

**Comment by theGreatHerrLebert** [comment]
Good catch! This issue is likely caused by setting the Scorer argument "annotate_matches" to False, which prevents detailed information about individual peaks matching the theoretical spectrum from being returned. I will address this in the next release by introducing a None check for the observed fragments in the respective function.

In the meantime, you can resolve the issue by setting the "annotate_matches" argument in the Scorer to True. However, keep in mind that re-scoring will only improve your search results if you provide informative features. The two most impactful features are predicted retention time and predicted fragment ion intensities. Without these, the re-scoring might actually yield worse results than the original hyper-score.

If you’re interested, I can provide a more detailed explanation of this process—it’s a fairly advanced topic involving machine learning. I’m also considering adding a small model for retention time prediction, similar to the regression model used in the original SAGE tool.

Best,

David

**Comment by daichengxin** [comment]
Thansk a lot. It works. That's great if sagepy can provide feature prediction functions or an API to input features?

**Comment by theGreatHerrLebert** [comment]
Thats a good idea, one could use the existing [KOINA](https://koina.wilhelmlab.org/) web service as an option to generate predicted features via the web, I might be adding this feature in the future.

**Comment by theGreatHerrLebert** [comment]
Hey there once again,

I've added two new notebooks—[property prediction](https://github.com/theGreatHerrLebert/sagepy/blob/main/sagepy/examples/property-prediction/property_prediction.ipynb)  and [re-scoring](https://github.com/theGreatHerrLebert/sagepy/blob/main/sagepy/examples/rescoring/rescoring.ipynb)—that demonstrate how to predict retention time, ion mobility (if applicable), and fragment ion intensity using another library we're currently developing. I may add support for external API calls at a later stage, but feel free to explore these notebooks in the meantime if you're interested.

The library for feature prediction, `imspy`, is also available via pip (`pip install imspy). Please note that it requires Python 3.11, as that's the only version TensorFlow currently supports.

Best regards,
David

**Comment by theGreatHerrLebert** [comment]
I was finally able to add the original SAGE based q-value calculations, check out [this notebook](https://github.com/theGreatHerrLebert/sagepy/blob/main/sagepy/examples/sage-fdr/FDRControl.ipynb).
